internal/util/reader.go

```diff
@@ -0,0 +1,107 @@
+package util
+
+import (
+	"compress/gzip"
+	"errors"
+	"io"
+)
+
+// PayloadReader is an implementation of io.Reader that reads bytes off the request body
+// optionally decompresses them, and has a limit attached.
+// If the limit is reached, an error will be returned and the underlying stream will be closed.
+//
+// Note: limit is applied to *both* compressed and uncompressed number of bytes.  This
+// protects us from potential zipbombs
+type PayloadReader struct {
+	IsGzipped bool
+	MaxBytes  int64
+
+	uncompressedBytesRead int64
+
+	wrappedBaseStream *byteCountingReader
+	stream            *io.LimitedReader
+}
+
+// NewReader creates a new reader
+func NewReader(r io.ReadCloser, isGzipped bool, maxBytes int64) (io.ReadCloser, error) {
+	// If this isn't compressed, and we don't want to limit the size, just
+	// return the original reader
+	if !isGzipped && maxBytes == 0 {
```

Comment:
To maintain backwards compatibility, if the new max payload size isn't specified,  it defaults to 0 which we will interpret to mean no restrictions on the size of the payload (the current behavior).

Because of this, we have a couple of early return chances here -- line 30 and 48. Because we are returning different types, the return type on line 26 is now an `io.ReadCloser` instead of the `PayloadReader` like one might expect.

This isn't exactly optimal. In a future release, when we don't mind breaking backwards compatibility, we won't have to deal with this 0 nonsense, and the logic can basically fall back through.

Alternatively, what do you think about making the config optional? If it isn't set, then we add no restriction, but once you specify a value, we enforce that limit, including 0.

Reply:
We have some existing payload size in the megabytes that will cause the service to send a 413. It seems like we could reasonably have a default limit that is some size larger than that.

Reply:
I like the idea that to maintain backwards compat, we allow the max payload size to not be set. 

If not set, then there is no restriction: old behavior.

If it is set, then it is restricted to whatever they choose.  We _could_ use -1 or 0 as a sentinel value to indicate "no restriction", but I think it would probably be better to enforce a positive number. That way you can be guaranteed that if you've configured this thing, you are protected to some extent.

Reply:
Update the code to use the new configtypes.OptBase2Bytes type. This lets it default to nothing being enforced, but once it is set, it must be > 0 so we can do some simple enforcing.